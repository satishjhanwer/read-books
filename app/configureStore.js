/**
 * Create the store with dynamic reducers
 */
import { applyMiddleware, createStore } from 'redux';
import { composeWithDevToolsDevelopmentOnly } from '@redux-devtools/extension';
import createSagaMiddleware from 'redux-saga';
import { createReduxHistoryContext } from 'redux-first-history';
import { persistStore } from 'reduxjs-toolkit-persist';
import { createBrowserHistory } from 'history';

import createReducer from './reducers';

const initialState = {};

const { createReduxHistory, routerMiddleware, routerReducer } = createReduxHistoryContext({
  history: createBrowserHistory(),
});

const reduxSagaMonitorOptions = {};
const sagaMiddleware = createSagaMiddleware(reduxSagaMonitorOptions);

export const store = createStore(
  createReducer(routerReducer),
  initialState,
  composeWithDevToolsDevelopmentOnly(applyMiddleware(...[sagaMiddleware, routerMiddleware])),
);

export const persistor = persistStore(store);
export const history = createReduxHistory(store);
