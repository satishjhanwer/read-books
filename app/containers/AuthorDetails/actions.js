/*
 *
 * AuthorDetails actions
 *
 */

import {
  AUTHOR_INFO_REQUEST,
  AUTHOR_INFO_SUCCESS,
  AUTHOR_INFO_FAILURE,
} from './constants';

/**
 *
 * @param {string} id Requested author id
 *
 * @return {object} An action object with a type of AUTHOR_INFO_REQUEST
 */
export function authorInfoRequest(id) {
  return {
    type: AUTHOR_INFO_REQUEST,
    id,
  };
}

/**
 *
 * @param {object} authorInfo Returned Author object based on author id
 *
 * @return {object} An action object with a type of AUTHOR_INFO_SUCCESS
 */
export function authorInfoSuccess(authorInfo) {
  return {
    type: AUTHOR_INFO_SUCCESS,
    authorInfo,
  };
}

/**
 *
 * @param {string} message error message
 *
 * @return {object} An action object with a type of AUTHOR_INFO_FAILURE
 */
export function authorInfoFailure(message) {
  return {
    type: AUTHOR_INFO_FAILURE,
    message,
  };
}
