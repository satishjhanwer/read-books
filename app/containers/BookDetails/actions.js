/*
 *
 * BookDetails actions
 *
 */

import {
  BOOK_INFO_REQUEST,
  BOOK_INFO_SUCCESS,
  BOOK_INFO_FAILURE,
} from './constants';

/**
 *
 * @param {string} id Requested book id
 *
 * @return {object} An action object with a type of BOOK_INFO_REQUEST
 */
export function requestBookInfo(id) {
  return { type: BOOK_INFO_REQUEST, id };
}

/**
 *
 * @param {object} book Returned Book object based on book id
 *
 * @return {object} An action object with a type of BOOK_INFO_SUCCESS
 */
export function requestBookInfoSuccess(book) {
  return { type: BOOK_INFO_SUCCESS, book };
}

/**
 *
 * @param {string} message error message
 *
 * @return {object} An action object with a type of BOOK_INFO_FAILURE
 */
export function requestBookInfoFailure(message) {
  return { type: BOOK_INFO_FAILURE, message };
}
