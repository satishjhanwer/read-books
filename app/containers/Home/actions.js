/*
 *
 * Home actions
 *
 */

import {
  BOOK_SEARCH_REQUEST,
  BOOK_SEARCH_SUCCESS,
  BOOK_SEARCH_FAILURE,
  CHANGE_SEARCH_TERM,
} from './constants';

export function changeSearchTerm(searchTerm) {
  return { type: CHANGE_SEARCH_TERM, searchTerm };
}

export function requestBookSearch(searchTerm) {
  return { type: BOOK_SEARCH_REQUEST, searchTerm };
}

export function requestBookSearchSuccess(bookList) {
  return { type: BOOK_SEARCH_SUCCESS, bookList };
}

export function requestBookSearchFailure(message) {
  return { type: BOOK_SEARCH_FAILURE, message };
}
