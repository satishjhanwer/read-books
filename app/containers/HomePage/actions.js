/*
 *
 * Home actions
 *
 */

import {
  BOOK_SEARCH_REQUEST,
  BOOK_SEARCH_SUCCESS,
  BOOK_SEARCH_FAILURE,
  CHANGE_SEARCH_TERM,
} from './constants';

/**
 *
 * @param {string} searchTerm The new text of input field
 *
 * @return {object} An action object with a type of CHANGE_SEARCH_TERM
 */
export function changeSearchTerm(searchTerm) {
  return { type: CHANGE_SEARCH_TERM, searchTerm };
}

/**
 *
 * @param {string} searchTerm The new text of input field
 *
 * @return {object} An action object with a type of BOOK_SEARCH_REQUEST
 */
export function requestBookSearch() {
  return { type: BOOK_SEARCH_REQUEST };
}

/**
 *
 * @param {Array} bookList List of books searched by search term
 *
 * @return {object} An action object with a type of BOOK_SEARCH_SUCCESS
 */
export function requestBookSearchSuccess(bookList) {
  return { type: BOOK_SEARCH_SUCCESS, bookList };
}

/**
 *
 * @param {string} message error message
 *
 * @return {object} An action object with a type of BOOK_SEARCH_FAILURE
 */
export function requestBookSearchFailure(message) {
  return { type: BOOK_SEARCH_FAILURE, message };
}
