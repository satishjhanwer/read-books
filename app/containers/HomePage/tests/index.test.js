/**
 * Test the HomePage
 */

import React from 'react';
import { render } from 'react-testing-library';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';

import { HomePage, mapDispatchToProps } from '../index';
import { changeSearchTerm, requestBookSearch } from '../actions';
import { store } from '../../../configureStore';

describe('<HomePage />', () => {
  it('should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <HomePage bookList={[]} />
        </IntlProvider>
      </Provider>,
    );
    expect(firstChild).toMatchSnapshot();
  });

  it('should fetch the book list on mount if a searchTerm exists', () => {
    const submitSpy = jest.fn();
    render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <HomePage searchTerm="Not Empty" onChangeSearchTerm={() => {}} onSubmitForm={submitSpy} />
        </IntlProvider>
      </Provider>,
    );
    expect(submitSpy).toBeDefined();
  });

  it('should not call onSubmitForm if searchTerm is an empty string', () => {
    const submitSpy = jest.fn();
    render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <HomePage onChangeSearchTerm={() => {}} onSubmitForm={submitSpy} />
        </IntlProvider>
      </Provider>,
    );
    expect(submitSpy).not.toHaveBeenCalled();
  });

  describe('mapDispatchToProps', () => {
    describe('onChangeSearchTerm', () => {
      it('should be injected', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        expect(result.onChangeSearchTerm).toBeDefined();
      });

      it('should dispatch changeSearchTerm when called', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        const searchTerm = 'harry';
        result.onChangeSearchTerm({ target: { value: searchTerm } });
        expect(dispatch).toHaveBeenCalledWith(changeSearchTerm(searchTerm));
      });
    });

    describe('onSubmitForm', () => {
      it('should be injected', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        expect(result.onSubmitForm).toBeDefined();
      });

      it('should dispatch requestBookSearch when called', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        result.onSubmitForm();
        expect(dispatch).toHaveBeenCalledWith(requestBookSearch());
      });

      it('should preventDefault if called with event', () => {
        const preventDefault = jest.fn();
        const result = mapDispatchToProps(() => {});
        const evt = { preventDefault };
        result.onSubmitForm(evt);
        expect(preventDefault).toHaveBeenCalledWith();
      });
    });
  });
});
