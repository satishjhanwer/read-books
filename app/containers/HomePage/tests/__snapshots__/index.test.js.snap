// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage /> should render and match the snapshot 1`] = `<br />`;
